crypto: update root certificates

bnoordhuis opened the pull request:
R=@iojs/crypto

Question: when are we going to drop the remaining 1024 bits certificates?

shigeki commented:
I tried this today but I found the tls access to `https://www.google.com/` was failed as `make test-internet` shows that
````
=== release test-http-https-default-ports ===
Path: internet/test-http-https-default-ports
events.js:141
      throw er; // Unhandled 'error' event
            ^
Error: certificate not trusted
    at Error (native)
    at TLSSocket.<anonymous> (_tls_wrap.js:918:38)
    at emitNone (events.js:67:13)
    at TLSSocket.emit (events.js:163:7)
    at TLSSocket._finishInit (_tls_wrap.js:506:8)
Command: out/Release/iojs /home/ohtsu/github/io.js/test/internet/test-http-https-default-ports.js
=== release test-tls-reuse-host-from-socket ===
Path: internet/test-tls-reuse-host-from-socket
events.js:141
      throw er; // Unhandled 'error' event
            ^
Error: certificate not trusted
    at Error (native)
    at TLSSocket.<anonymous> (_tls_wrap.js:918:38)
    at emitNone (events.js:67:13)
    at TLSSocket.emit (events.js:163:7)
    at TLSSocket._finishInit (_tls_wrap.js:506:8)
Command: out/Release/iojs /home/ohtsu/github/io.js/test/internet/test-tls-reuse-host-from-socket.js
[00:06|% 100|+   8|-   2]: Done
````
I'm now investing it.

bnoordhuis commented:
Good catch, @shigeki.  I think it has to do with one of the Equifax certificates that are now removed.

bnoordhuis commented:
Okay, so adding the Equifax Secure CA back in fixes it:

```diff
diff --git a/src/node_root_certs.h b/src/node_root_certs.h
index e650ac3..7a13a02 100644
--- a/src/node_root_certs.h
+++ b/src/node_root_certs.h
@@ -120,6 +120,25 @@
 "UPMxxY8BqHTr9Xgn2uf3ZkPznoM+IKrDNWCRzg==\n"
 "-----END CERTIFICATE-----\n",
 
+/* Equifax Secure CA */
+"-----BEGIN CERTIFICATE-----\n"
+"MIIDIDCCAomgAwIBAgIENd70zzANBgkqhkiG9w0BAQUFADBOMQswCQYDVQQGEwJVUzEQMA4G\n"
+"A1UEChMHRXF1aWZheDEtMCsGA1UECxMkRXF1aWZheCBTZWN1cmUgQ2VydGlmaWNhdGUgQXV0\n"
+"aG9yaXR5MB4XDTk4MDgyMjE2NDE1MVoXDTE4MDgyMjE2NDE1MVowTjELMAkGA1UEBhMCVVMx\n"
+"EDAOBgNVBAoTB0VxdWlmYXgxLTArBgNVBAsTJEVxdWlmYXggU2VjdXJlIENlcnRpZmljYXRl\n"
+"IEF1dGhvcml0eTCBnzANBgkqhkiG9w0BAQEFAAOBjQAwgYkCgYEAwV2xWGcIYu6gmi0fCG2R\n"
+"FGiYCh7+2gRvE4RiIcPRfM6fBeC4AfBONOziipUEZKzxa1NfBbPLZ4C/QgKO/t0BCezhABRP\n"
+"/PvwDN1Dulsr4R+AcJkVV5MW8Q+XarfCaCMczE1ZMKxRHjuvK9buY0V7xdlfUNLjUA86iOe/\n"
+"FP3gx7kCAwEAAaOCAQkwggEFMHAGA1UdHwRpMGcwZaBjoGGkXzBdMQswCQYDVQQGEwJVUzEQ\n"
+"MA4GA1UEChMHRXF1aWZheDEtMCsGA1UECxMkRXF1aWZheCBTZWN1cmUgQ2VydGlmaWNhdGUg\n"
+"QXV0aG9yaXR5MQ0wCwYDVQQDEwRDUkwxMBoGA1UdEAQTMBGBDzIwMTgwODIyMTY0MTUxWjAL\n"
+"BgNVHQ8EBAMCAQYwHwYDVR0jBBgwFoAUSOZo+SvSspXXR9gjIBBPM5iQn9QwHQYDVR0OBBYE\n"
+"FEjmaPkr0rKV10fYIyAQTzOYkJ/UMAwGA1UdEwQFMAMBAf8wGgYJKoZIhvZ9B0EABA0wCxsF\n"
+"VjMuMGMDAgbAMA0GCSqGSIb3DQEBBQUAA4GBAFjOKer89961zgK5F7WF0bnj4JXMJTENAKaS\n"
+"bn+2kmOeUJXRmm/kEd5jhW6Y7qj/WsjTVbJmcVfewCHrPSqnI0kBBIZCe/zuf6IWUrVnZ9NA\n"
+"2zsmWLIodz2uFHdh1voqZiegDfqnc1zqcPGUIWVEX/r87yloqaKHee9570+sB3c4\n"
+"-----END CERTIFICATE-----\n",
+
 /* Entrust.net Premium 2048 Secure Server CA */
 "-----BEGIN CERTIFICATE-----\n"
 "MIIEKjCCAxKgAwIBAgIEOGPe+DANBgkqhkiG9w0BAQUFADCBtDEUMBIGA1UEChMLRW50cnVz\n"
```

I'll do some digging as to why it was removed.

bnoordhuis commented:
I suspect the cause may be more benign and we simply need to update the mk-ca-bundle.pl script.

shigeki commented:
Yes, I think so. Some labels in certdata.txt are changed. It need to be updated from the latest script in the curl distribution.

bnoordhuis commented:
I updated the mk-ca-bundle.pl script but we may be in a pickle.  The reason the Equifax certificate is skipped is because of this (from its entry in certdata.txt):

```
CKA_TRUST_SERVER_AUTH CK_TRUST CKT_NSS_MUST_VERIFY_TRUST
CKA_TRUST_EMAIL_PROTECTION CK_TRUST CKT_NSS_TRUSTED_DELEGATOR
CKA_TRUST_CODE_SIGNING CK_TRUST CKT_NSS_MUST_VERIFY_TRUST
CKA_TRUST_STEP_UP_APPROVED CK_BBOOL CK_FALSE
```

That is, it's trusted as a CA for EMAIL_PROTECTION but not SERVER_AUTH (which is the default filter.)

When I run the script with `-p EMAIL_PROTECTION,SERVER_AUTH:TRUSTED_DELEGATOR`, the certificate gets included, but so are a number of others that I'm not sure we want to include:

```
AC Ra\xC3\xADz Certic\xC3\xA1mara S.A.
ComSign CA
Digital Signature Trust Co. Global CA 1
Digital Signature Trust Co. Global CA 3
Equifax Secure CA
Equifax Secure Global eBusiness CA
Equifax Secure eBusiness CA 1
NetLock Business (Class B) Root
NetLock Express (Class C) Root
NetLock Qualified (Class QA) Root
S-TRUST Authentication and Encryption Root CA 2005 PN
S-TRUST Universal Root CA
SG TRUST SERVICES RACINE
Sonera Class 1 Root CA
SwissSign Platinum CA - G2
TC TrustCenter Class 3 CA II
UTN USERFirst Email Root CA
Verisign Class 1 Public Primary Certification Authority
Verisign Class 1 Public Primary Certification Authority - G2
Verisign Class 1 Public Primary Certification Authority - G3
Verisign Class 2 Public Primary Certification Authority - G2
Verisign Class 2 Public Primary Certification Authority - G3
Verisign Class 3 Public Primary Certification Authority
Verisign Class 3 Public Primary Certification Authority
Verisign Class 3 Public Primary Certification Authority - G2
```

Suggestions?

Fishrock123 commented on tools/mk-ca-bundle.pl:
```diff
@@ -34,35 +34,85 @@
+use List::Util;
+use Text::Wrap;
 
-my $url = 'http://mxr.mozilla.org/mozilla/source/security/nss/lib/ckfw/builtins/certdata.txt?raw=1';
```
There was some concern in https://github.com/joyent/node/pull/13860 that this wasn't `https`. Does this supersede that entirely?

bnoordhuis replied:
I don't know if 'supersede' is the right word because our version of the script doesn't download the file.  It's up to the committer to download it to tools/certdata.txt.

shigeki commented:
I found that the issue caused by the alt cert chain problem in openssl. The certificate chain of www.google.com shows it has a cross root cert betwenn Equifax Secure Certificate Authority and GeoTrust Global CA at 2. 
The openssl client using  a ca file which has no Equifax CA shows as below with and without alt chain.
````
$ ./openssl s_client -no_alt_chains -connect www.google.com:443 -CAfile ~/ca-bundle.crt
WARNING: can't open config file: /usr/local/ssl/openssl.cnf
CONNECTED(00000003)
depth=2 C = US, O = GeoTrust Inc., CN = GeoTrust Global CA
verify error:num=20:unable to get local issuer certificate
---
Certificate chain
 0 s:/C=US/ST=California/L=Mountain View/O=Google Inc/CN=www.google.com
   i:/C=US/O=Google Inc/CN=Google Internet Authority G2
 1 s:/C=US/O=Google Inc/CN=Google Internet Authority G2
   i:/C=US/O=GeoTrust Inc./CN=GeoTrust Global CA
 2 s:/C=US/O=GeoTrust Inc./CN=GeoTrust Global CA
   i:/C=US/O=Equifax/OU=Equifax Secure Certificate Authority
---

$ ./openssl s_client -connect www.google.com:443 -CAfile ~/ca-bundle.crt
WARNING: can't open config file: /usr/local/ssl/openssl.cnf
CONNECTED(00000003)
depth=2 C = US, O = GeoTrust Inc., CN = GeoTrust Global CA
verify return:1
depth=1 C = US, O = Google Inc, CN = Google Internet Authority G2
verify return:1
depth=0 C = US, ST = California, L = Mountain View, O = Google Inc, CN = www.google.com
verify return:1
---
Certificate chain
 0 s:/C=US/ST=California/L=Mountain View/O=Google Inc/CN=www.google.com
   i:/C=US/O=Google Inc/CN=Google Internet Authority G2
 1 s:/C=US/O=Google Inc/CN=Google Internet Authority G2
   i:/C=US/O=GeoTrust Inc./CN=GeoTrust Global CA
 2 s:/C=US/O=GeoTrust Inc./CN=GeoTrust Global CA
   i:/C=US/O=Equifax/OU=Equifax Secure Certificate Authority
---
````
Without alt chains, connecting to google.com are not verified.

https://github.com/shigeki/io.js/tree/upgrade_openssl102_and_upgrade_rootCA is my branch update rootCA with openssl-1.0.2 with alt cert chain patch. `make test-internet` works fine without Equifax Secure CA as
````
ohtsu@spdy:~/github/shigeki/io.js$ grep 'Equifax Secure CA' src/node_root_certs.h
ohtsu@spdy:~/github/shigeki/io.js$ make test-internet
make -C out BUILDTYPE=Release V=1
make[1]: Entering directory `/home/ohtsu/github/shigeki/io.js/out'
make[1]: Nothing to be done for `all'.
make[1]: Leaving directory `/home/ohtsu/github/shigeki/io.js/out'
ln -fs out/Release/iojs iojs
/usr/bin/python tools/test.py internet
[00:05|% 100|+  10|-   0]: Done
````
I will work on openssl-1.0.2a tomorrow. I think it is better to update root CA after that.

bnoordhuis commented:
@shigeki Apropos openssl 1.0.2: I know we discussed bundling nasm or yasm but I looked at it this weekend and I think it may be possible to simply post-process the generated .s files to make them acceptable to GNU as.

The only requirement is that gas / binutils is new enough to know about AVX-256.  That probably means 2.23 or 2.24 as the minimum but that seems acceptable to me.

shigeki commented:
@bnoordhuis nasm or yasm requirement is for building on Win32. IIRC, the current build requirements of iojs on Win32 are only python and Visual Studio. Are you planning to have an additional build requirement to install gas/binutils on Win32 or put them into `deps` to build them within gyp?

indutny commented:
@bnoordhuis @shigeki I suggest backporting:

da084a5ec6
15dba5be6a
25690b7f5f
fa7b01115b

From OpenSSL master once we'll switch to 1.0.2, this should solve 1024bit CA problem once and for all.

indutny commented:
btw, LGTM! Good job!

shigeki commented:
@indutny I backported the first and second one in https://github.com/shigeki/io.js/commit/90368ee2084d82729a54ebc3ed98e3d160b47d10 and https://github.com/shigeki/io.js/commit/39128d65e2ac27543bc23dae23659d49b6419e56 but not the rest of two. Do you need the apps and doc patches too?

indutny commented:
oh @shigeki ! You are awesome.

shigeki commented:
I planed to work this again in this week with the latest certdata.txt and deprecating 1024bit root certs. But mozilla has found an issue of Equifax 1024-bit roots as https://bugzilla.mozilla.org/show_bug.cgi?id=1155279 .   Wait and see it for a moment.

bnoordhuis commented:
@nodejs/crypto I updated to the latest certdata.txt, PTAL.  The target branch is wrong now because of the switch to master but it's the last four commits.

@shigeki Do I understand correctly that bnoordhuis/io.js@a87651a ("crypto: add back deprecated 1024 bits root certs") is no longer necessary?  I've dropped it for now but I can add it again if necessary.

shigeki commented:
@bnoordhuis Is it better to open a new PR against current master? This PR is something strange that contains 173 commits. 
 As for 1024 bits root certs, yes we no longer need to add some 1024 bits certs for websites that has a cross root cert. We can just follow mozilla's certdata.txt, where one 1024 bits root CA is re-enabled as in https://bugzilla.mozilla.org/show_bug.cgi?id=1155279.

bnoordhuis commented:
I wanted to keep the PR URL the same, to keep a backlink to the conversation from March.  I guess it did become a little befuddled though.  I'll file a new PR.
